File: python/tests/test_order_tif.py
import datetime as dt

import regimeflow as rf


def test_order_expire_at_roundtrip():
    expire_at = dt.datetime(2025, 1, 1, 12, 0, 0)
    order = rf.Order(
        symbol="AAPL",
        side=rf.OrderSide.BUY,
        type=rf.OrderType.MARKET,
        quantity=1.0,
        tif=rf.TimeInForce.GTD,
        expire_at=expire_at,
    )
    assert order.expire_at is not None
    assert order.expire_at.year == 2025

File: tests/unit/test_order_tif.cpp
#include "regimeflow/engine/order_manager.h"

#include <gtest/gtest.h>

using regimeflow::Timestamp;
using regimeflow::engine::Order;
using regimeflow::engine::OrderManager;
using regimeflow::engine::OrderSide;
using regimeflow::engine::OrderType;
using regimeflow::engine::TimeInForce;

TEST(OrderTifTest, GtdRequiresExpireAt) {
    OrderManager manager;
    Order order = Order::market(regimeflow::SymbolRegistry::instance().intern("AAA"),
                                OrderSide::Buy,
                                10.0);
    order.tif = TimeInForce::GTD;
    auto result = manager.submit_order(order);
    EXPECT_TRUE(result.is_err());
}

TEST(OrderTifTest, GtdAllowsExpireAt) {
    OrderManager manager;
    Order order = Order::market(regimeflow::SymbolRegistry::instance().intern("AAA"),
                                OrderSide::Buy,
                                10.0);
    order.tif = TimeInForce::GTD;
    order.expire_at = Timestamp::now();
    auto result = manager.submit_order(order);
    EXPECT_TRUE(result.is_ok());
}
